mock/src/main/java/feign/mock/RequestKey.java

```diff
@@ -0,0 +1,173 @@
+        }
+
+        public Builder body(String body) {
+            return body(body.getBytes(UTF_8));
+        }
+
+        public Builder body(byte[] body) {
+            this.body = body;
+            return this;
+        }
+
+        public RequestKey build() {
+            return new RequestKey(this);
+        }
+
+    }
+
+    public static Builder builder(HttpMethod method, String url) {
+        return new Builder(method, url);
+    }
+
+    public static RequestKey create(Request request) {
+        return new RequestKey(request);
+    }
+
+    private static String buildUrl(Request request) {
+        try {
+            return URLDecoder.decode(request.url(), Util.UTF_8.name());
+        } catch (final UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final HttpMethod method;
+
+    private final String url;
+
+    private final Map<String, Collection<String>> headers;
+
+    private final Charset charset;
+
+    private final byte[] body;
+
+    private RequestKey(Builder builder) {
+        this.method = builder.method;
+        this.url = builder.url;
+        this.headers = builder.headers;
+        this.charset = builder.charset;
+        this.body = builder.body;
+    }
+
+    private RequestKey(Request request) {
+        this.method = HttpMethod.valueOf(request.method());
+        this.url = buildUrl(request);
+        this.headers = request.headers();
+        this.charset = request.charset();
+        this.body = request.body();
+    }
+
+    public HttpMethod getMethod() {
+        return method;
+    }
+
+    public String getUrl() {
+        return url;
+    }
+
+    public Map<String, Collection<String>> getHeaders() {
+        return headers;
+    }
+
+    public Charset getCharset() {
+        return charset;
+    }
+
+    public byte[] getBody() {
+        return body;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((method == null) ? 0 : method.hashCode());
+        result = prime * result + ((url == null) ? 0 : url.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        final RequestKey other = (RequestKey) obj;
+        if (method != other.method) return false;
+        if (url == null) {
+            if (other.url != null) return false;
+        } else if (!url.equals(other.url)) return false;
+        return true;
+    }
+
+    public boolean equalsExtended(Object obj) {
+        if (equals(obj)) {
+            RequestKey other = (RequestKey) obj;
+            boolean headersEqual = other.headers == null || headers == null || headers.equals(other.headers);
+            boolean charsetEqual = other.charset == null || charset == null || charset.equals(other.charset);
```

Comment:
Does this intentionally consider `other.chaset == "UTF8"` and `charset == null` to be equal?